### ISSUE TITLE
Handling invalid Polygon geometries during load_table_from_dataframe

### ISSUE THREAD

**Issue by mradamcox** [googleapis/python-bigquery#1638]
Hello! I'm working on a spatial data ETL pipeline into BigQuery, and am having trouble with how to approach data cleanup/validation for certain geometries. I am loading shapefiles into GeoPandas dataframes and then using load_table_from_dataframe. The first file I have worked fine, but the next fails with

```
google.api_core.exceptions.BadRequest: 400 Error while reading data, error message: Invalid geography value, error: Polygon's first loop must be shell. It is nested in loop 2; in polygon 2; in WKB geography; in column 'geom'
```

Thing is, I've used a few validation methods, like running "Fix Geometries" in QGIS or applying [shapely operations](https://shapely.readthedocs.io/en/stable/reference/shapely.Polygon.html) like normalize() and make_valid(), but nothing is helping with this error.

Ideally, I'd be able to load the table but ignore these error rows and get an output message saying which rows failed, because as it is, I'm not even sure which particular polygon is failing.

Any help appreciated! I've also seen there are spatial functions available in BigQuery... perhaps these could be used to fix the geometries, but I don't see how to implement them in Python without first loading the data to BigQuery.

**Comment by meredithslota** [comment]
Hi @mradamcox — this documentation covers the supported types of geospatial data: https://cloud.google.com/bigquery/docs/geospatial-data Typically this error is due to problems in the data; here's a similar issue that has some suggested solutions: https://stackoverflow.com/questions/72577936/badrequest-load-geodataframe-to-bigquery-python

From that post, it seems like common problems are real invalid data (sounds like probably not if you are running validation methods?) or incorrect polygon loop orientation. This library doesn't handle fixing those issues specifically, but you could use the `skip_invalid_rows` option to upload only the rows without this issue: https://github.com/googleapis/python-bigquery/blob/main/google/cloud/bigquery/client.py#L3566 (if you are using a supported and valid geospatial datatype).

**Comment by mradamcox** [comment]
Hi @meredithslota, thank you for your response. I had been validating the geometries as best I could, ring orientation for example, but this particular "first loop" issue wasn't getting fixed by any of the methods I tried, and I wasn't able to iterate the data ahead of time to check for it.

However, I also found that stack exchange question and it's actually the one that put me on the right pat--I ended up loading the geometries to JSON to avoid this loop order problem.

I was planning to use that exact same `skip_invalid_rows` flag, but I found that those data streaming load methods aren't available under the free tier (which we are currently using). Ultimately, I did get it to work by "faking" a newline-delimited GeoJSON file and loading from json file.

I also realized that I wasn't wrapping the `load_job.result()` in a `try/except` block, and once I did that I was able to get some better error messages.

```
    str_data = "\n".join(rows)
    data_as_file = io.StringIO(str_data)

    full_table_id = f"{project_id}.{schema['bq_dataset_name']}.{schema['bq_table_name']}"
    table = client.get_table(full_table_id)

    load_job_config = LoadJobConfig(
        source_format="NEWLINE_DELIMITED_JSON"
    )
    load_job = client.load_table_from_file(
        file_obj=data_as_file,
        destination=table,
        job_config=load_job_config
    )
    print("Running job now...")
    try:
        load_job.result()
    except Exception as e:
        print(f"Errors encountered: {len(load_job.errors)}")
        for error in load_job.errors:
            print(error)
        raise e
    print(f"Output rows: {load_job.output_rows}")
```

All that to say, I _think_ I have this handled, but appreciate your response very much. It would be nice if this client could also perform geometry checks discretely, independent of the load job process, but like I mentioned, I don't think I was properly handling the exceptions at first and may have been able to get further along if I had.
